up42/auth.py

```diff
@@ -100,66 +111,112 @@ def __init__(
+
+        if project_id:
+            self.project_id = project_id
+
+        config = self._choose_credential_source(cfg_file, credentials_dict)
+        self._set_credentials(source=config)
+
         if self.authenticate:
-            self._find_credentials()
             self._get_token()
-            self._get_workspace()
+            self._get_user_id()
             logger.info("Authentication with UP42 successful!")
 
     def __repr__(self):
         env_string = f" ,{self.env}" if self.env != "com" else ""
-        return f"UP42ProjectAuth(project_id={self.project_id}{env_string})"
-
-    def _find_credentials(self) -> None:
-        """
-        Sources the project credentials from a provided config file, error handling
-        if no credentials are provided in arguments or config file.
-        """
-        if self.project_id is None or self.project_api_key is None:
-            if self.cfg_file is None:
-                raise ValueError("Provide project_id and project_api_key via arguments or config file!")
+        return f"UP42Auth(id={self._credentials_id}{env_string})"
 
-            # Source credentials from config file.
+    def _choose_credential_source(self, cfg_file: Union[str, Path, None], kwargs: dict) -> dict:
+        config = {}
+        if cfg_file:
             try:
-                with open(self.cfg_file) as src:
-                    config = json.load(src)
-                    try:
-                        self.project_id = config["project_id"]
-                        self.project_api_key = config["project_api_key"]
-                    except KeyError as e:
-                        raise ValueError(
-                            "Provided config file does not contain project_id and " "project_api_key!"
-                        ) from e
+                with open(cfg_file, encoding="utf-8") as src:
+                    config.update(json.load(src))
                 logger.info("Got credentials from config file.")
             except FileNotFoundError as e:
                 raise ValueError("Selected config file does not exist!") from e
 
-        elif all(v is not None for v in [self.cfg_file, self.project_id, self.project_api_key]):
-            logger.info(
-                "Credentials are provided via arguments and config file, " "now using the argument credentials."
-            )
+            if set(config.keys()).intersection(set(kwargs.keys())):
+                raise ValueError("Credentials must be provided either via config file or arguments, but not both")
+        return config if config else kwargs
+
+    def _set_credentials(self, source: dict):
+        schemas = {
+            "user": ("username", "password"),
+            "project": ("project_id", "project_api_key"),
+        }
+        token_retrievers = {
+            "user": self._get_token_account_based,
+            "project": self._get_token_project_based,
+        }
+
+        if set(source.keys()) == {"project_id"}:
```

Comment:
Why is this needed?

Reply:
Because of this use case. The user Initializes the project by only providing the project_id (not the project_api_key).
The suggested workflow in set_credentials is that if there is not pair (username, password) or (project_id, project_api_key) it raises an error. Therefore, by adding the case of only adding project_id and return before authentication, makes this use case available.

```
def test_should_initialize_project_with_project_id(auth_mock, requests_mock):
    up42.authenticate(
        project_id=PROJECT_ID,
        authenticate=False,
    )
```